runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/PTransformMatchersTest.java

```diff
@@ -270,4 +276,107 @@ public void parDoMultiWithTimers() {
     assertThat(PTransformMatchers.splittableParDoSingle().matches(parDoApplication), is(false));
     assertThat(PTransformMatchers.stateOrTimerParDoSingle().matches(parDoApplication), is(false));
   }
+
+  @Test
+  public void parDoWithFnTypeWithMatchingType() {
+    DoFn<Object, Object> fn = new DoFn<Object, Object>() {
+      @ProcessElement
+      public void process(ProcessContext ctxt) {
+      }
+    };
+    AppliedPTransform<?, ?, ?> parDoSingle = getAppliedTransform(ParDo.of(fn));
+    AppliedPTransform<?, ?, ?> parDoMulti =
+        getAppliedTransform(
+            ParDo.of(fn).withOutputTags(new TupleTag<Object>(), TupleTagList.empty()));
+
+    PTransformMatcher matcher = PTransformMatchers.parDoWithFnType(fn.getClass());
+    assertThat(matcher.matches(parDoSingle), is(true));
+    assertThat(matcher.matches(parDoMulti), is(true));
+  }
+
+  @Test
+  public void parDoWithFnTypeWithNoMatch() {
+    DoFn<Object, Object> fn = new DoFn<Object, Object>() {
+      @ProcessElement
+      public void process(ProcessContext ctxt) {
+      }
+    };
+    AppliedPTransform<?, ?, ?> parDoSingle = getAppliedTransform(ParDo.of(fn));
+    AppliedPTransform<?, ?, ?> parDoMulti =
+        getAppliedTransform(
+            ParDo.of(fn).withOutputTags(new TupleTag<Object>(), TupleTagList.empty()));
+
+    PTransformMatcher matcher = PTransformMatchers.parDoWithFnType(doFnWithState.getClass());
+    assertThat(matcher.matches(parDoSingle), is(false));
```

Comment:
Eventually we might need a way for a `ParDo` to exist that does not have a valid Java DoFn _or_ have a `PTransform` with a URN of `urn:beam:transform:pardo` that is not an instance of `ParDo.Bound`, etc. Just a heads up. We aren't there yet, and the problem will be its own forcing function when we get there.